### PR TITLE
Fix the bugs for AzKeyStore

### PR DESCRIPTION
**Pull request by msJinLei**
## Description

* Enabled credential to be found only by applicationId while tenant was not matched when accquire token. [#20484]
* When Az.Accounts ran in parallel, the waiters were allowed to wait infinitely to avoid throw exception in automation enviroment. [#20455]
* Used Lazy load for AzKeyStore.
* Used update on change mechanism for AzKeyStore and remove `Flush` interface.

## Checklist

- [ ] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Review by isra-fel — CHANGES_REQUESTED** [review]
I'll review the remaining code afterwards

File: src/Accounts/Authentication/KeyStore/StorageHelper.cs [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+            if (keystore == null)
+            {
+                throw new ArgumentNullException(nameof(keystore));
+            }
+            keystore.SetOnUpdate(null);
+        }
+
+        public void LoadFromCachedStorage(IKeyStore keystore)
+        {
+            LogMessage(EventLogLevel.Verbose, $"Before access\nAcquiring lock for token cache");
```
Comment (isra-fel) [inline src/Accounts/Authentication/KeyStore/StorageHelper.cs:174]: token cache

File: src/Accounts/Authentication/KeyStore/StorageHelper.cs [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                try
+                {
+                    data = args.KeyStore.Serialize();
+                }
+                catch (Exception e)
+                {
+                    LogMessage(EventLogLevel.Error, $"An exception was encountered while serializing the {nameof(StorageHelper)} : {e}");
+                    LogMessage(EventLogLevel.Error, $"No data found in the store, clearing the cache in memory.");
+
+                    PersistanceStore.Clear();
```
Comment (isra-fel) [inline src/Accounts/Authentication/KeyStore/StorageHelper.cs:225]: Clear calls WriteToCachedStorage recursively.

File: tools/Test/SmokeTest/InstallAzModules.ps1 [hunk 3]
```diff
@@ -42,6 +42,12 @@ Register-Gallery $gallery $localRepoLocation
 
 Write-Host "Installing Az..."
 Install-Module -Name Az -Repository $gallery -Scope CurrentUser -AllowClobber -Force 
+
+$file = Get-ChildItem $localRepoLocation | Where-Object {$_.Name -like "ThreadJob*"}
+if ($file -ne $null) {
```
Comment (isra-fel) [inline tools/Test/SmokeTest/InstallAzModules.ps1:47]:
```suggestion
if ($file -ne $null and threadjob is not already installed) {
```
low priority

File: tools/Test/SmokeTest/RmCoreSmokeTests.ps1 [hunk 4]
```diff
@@ -184,6 +184,28 @@ $generalCommands = @(
         };
         Retry = 0; # no need to retry
     }
+    @{
+        Name = "Import Az.Accounts in Parallel Thread Job";
+
+        Command = {
+            #if ($null -ne $env:SYSTEM_DEFINITIONID -or $null -ne $env:Release_DefinitionId) {
```
Comment (isra-fel) [inline tools/Test/SmokeTest/RmCoreSmokeTests.ps1:191]: delete

File: tools/Test/SmokeTest/RmCoreSmokeTests.ps1 [hunk 5]
```diff
@@ -164,7 +164,7 @@ $resourceTestCommands = @(
 
 $generalCommands = @(
     @{
-        Name = "Import Az.Accounts in Parallel";
+        Name = "Import Az.Accounts in Parallel Job";
```
Comment (isra-fel) [inline tools/Test/SmokeTest/RmCoreSmokeTests.ps1:167]:
```suggestion
        Name = "Import Az.Accounts in Parallel (Process)";
```

File: tools/Test/SmokeTest/RmCoreSmokeTests.ps1 [hunk 6]
```diff
@@ -184,6 +184,28 @@ $generalCommands = @(
         };
         Retry = 0; # no need to retry
     }
+    @{
+        Name = "Import Az.Accounts in Parallel Thread Job";
```
Comment (isra-fel) [inline tools/Test/SmokeTest/RmCoreSmokeTests.ps1:188]:
```suggestion
        Name = "Import Az.Accounts in Parallel (Thread)";
```

**Review by isra-fel — CHANGES_REQUESTED** [review]
- When we enable/disable token cache persistence, we should enable/disable flushing key store as well

File: src/Accounts/Authentication/KeyStore/StorageHelper.cs (hunk 2 above)
Comment (msJinLei) [reply src/Accounts/Authentication/KeyStore/StorageHelper.cs:225]: Sorry for the mistake. PersistanceStore clear won't call WriteToCachedStorage recsusively. It's InMemoryKeyStore clear will do.

File: src/Accounts/Authentication/KeyStore/ServicePrincipalKey.cs [hunk 7]
```diff
@@ -54,5 +54,14 @@ public override bool Equals(object obj)
             }
             return false;
         }
+
+        public override bool BeEquivalent(object obj)
+        {
+            if (obj is ServicePrincipalKey other)
+            {
+                return this.name == other.name && this.appId == other.appId;
```
[the remark this answers is not in the document]
Comment (msJinLei) [reply src/Accounts/Authentication/KeyStore/ServicePrincipalKey.cs:62]: Add comments to interface.

**Review by isra-fel — APPROVED** [review]

**Pull request merged**